[GPU Process] [Filters] Transfer the ownership of the FilterImage buffers from GPUProcess to WebProcess

shallawa opened the pull request:
#### 2166784c7ea6086a359ebd1b6df846ed3040cdfe
<pre>
[GPU Process] [Filters] Transfer the ownership of the FilterImage buffers from GPUProcess to WebProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=240811">https://bugs.webkit.org/show_bug.cgi?id=240811</a>
&lt;rdar://problem/94118219&gt;

Reviewed by Chris Dumez.

When creating an ImageBuffer or a PixelBuffer for FilterImage in GPUProcess, we
need to transfer the ownership of this memory to WebProcess. For this purpose
we are going to call mach_memory_entry_ownership() and passing the handle of the
SharedMemomry which backs these buffers and the ProcessIdentity of the WebProcess.

* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
(WebKit::ImageBufferShareableAllocator::ImageBufferShareableAllocator):
(WebKit::ImageBufferShareableAllocator::transferMemoryOwnership const):
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::transferMemoryOwnership):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Platform/MemoryLedger.h: Copied from Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.h.
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::transferMemoryOwnership):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/251798@main">https://commits.webkit.org/251798@main</a>
</pre>

jyavenard reviewed (commented):
LGTM.

Unfortunately, I don't have committer rights yet.

But r=me :)

Hopefully the kernel API will be available soon, allowing immediate ownership change instead of having to send it over the content process.

cdumez commented on Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
```diff
@@ -95,17 +96,23 @@ SharedMemory::Handle::~Handle()
 }
 
-void SharedMemory::Handle::takeOwnershipOfMemory(MemoryLedger memoryLedger) const
+void SharedMemory::Handle::takeOwnershipOfMemory(const ProcessIdentity& processIdentity, MemoryLedger memoryLedger) const
```
I think we should introduce a new function called setOwnershipOfMemory(const ProcessIdentity&, MemoryLedger) instead.

Having something called "takeOwnershipOfMemory" but passing the identity of another process in parameter is super confusing.

shallawa replied:
The new function SharedMemory::setOwnershipOfMemory() was added.

cdumez commented on Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
```diff
@@ -95,17 +96,23 @@ SharedMemory::Handle::~Handle()
     if (!m_port)
         return;
 
+#if HAVE(TASK_IDENTITY_TOKEN)
```
This is not going to suffice since support for mach_memory_entry_ownership() taking in an identity token is much more recent than generic identity token support (HAVE(TASK_IDENTITY_TOKEN)).

shallawa replied:
The macro HAVE_MACH_MEMORY_ENTRY_OWNERSHIP_IDENTITY_TOKEN_SUPPORT was added for the new behavior of mach_memory_entry_ownership().

cdumez reviewed (approved):
LGTM

shallawa committed:
[GPU Process] [Filters] Transfer the ownership of the FilterImage buffers from GPUProcess to WebProcess
https://bugs.webkit.org/show_bug.cgi?id=240811
<rdar://problem/94118219>

Reviewed by Chris Dumez.

When creating an ImageBuffer or a PixelBuffer for FilterImage in GPUProcess, we
need to transfer the ownership of this memory to WebProcess. For this purpose
we are going to call mach_memory_entry_ownership() and passing the handle of the
SharedMemomry which backs these buffers and the ProcessIdentity of the WebProcess.

* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
(WebKit::ImageBufferShareableAllocator::ImageBufferShareableAllocator):
(WebKit::ImageBufferShareableAllocator::transferMemoryOwnership const):
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::transferMemoryOwnership):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Platform/MemoryLedger.h: Copied from Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.h.
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::transferMemoryOwnership):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:

Canonical link: https://commits.webkit.org/251798@main
